gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,278 @@
+            depth_ranges[i, 1] = np.percentile(depths[i], MAX_DEPTH_PERCENTILE)
+
+        return src_views_dict, depth_ranges
+
+    def __len__(self) -> int:
+        """Returns the number of reference views in the dataset. A depth map will be estimated for each.
+
+        In the PatchmatchNet workflow, each input image will be treated as a reference view once, and each input
+        image corresponds to a estimated camera pose.
+
+        Returns:
+            the length of the dataset.
+        """
+        return self._num_valid_cameras
+
+    def __getitem__(self, index: int) -> Dict[str, Any]:
+        """Get inference data to PatchmatchNet
+        Produces test data from one inference view.
+        Returns meta data including ids and depth range, input images(_num_views images, the first images is the
+        reference image) and projection matrices, as well as the output filename
+
+        Args:
+            index: index of yield test data, the reference image ID of the test data the patchmatchnet index (from 0 to
+            number_of_views - 1), not the camera index in the sfm_result
+
+        Returns:
+            Dictionary containing:
+                "idx" test image index: int
+                "imgs" source and reference images: (num_views, image_channel, image_h, image_w)
+                "proj_matrices" projection matrices: (num_views, 4, 4)
+                "depth_min" minimum depth: int
+                "depth_max" maximum depth: int
+                "filename" output filename pattern: string
+        """
+        ref_idx = self._patchmatchnet_idx_to_camera_idx[index]
+        src_idxs = [self._patchmatchnet_idx_to_camera_idx[src_index] for src_index in self._src_views_dict[index]]
+
+        imgs: List[List[np.ndarray]] = [[] for _ in range(self._num_stages)]
+        proj_mats: List[List[np.ndarray]] = [[] for _ in range(self._num_stages)]
+
+        for i in [ref_idx] + src_idxs:
+            img = self._images[i].value_array
+            np_img = np.array(img, dtype=np.float32) / 255.0
+            np_img = cv2.resize(np_img, (self._w, self._h), interpolation=cv2.INTER_LINEAR)
+            # Crop the image from the upper left corner, instead of from the center
+            #   to align with the 2D coordinates in track's measurements.
+            np_img = np_img[: self._cropped_h, : self._cropped_w, :]
+
+            intrinsics = self._sfm_result.get_camera(i).calibration().K()
+            cTw = self._sfm_result.get_camera(i).pose().inverse().matrix()
+            # In the multi-scale feature extraction, there are NUM_PATCHMATCHNET_STAGES stages.
+            #   Resize the image to scales in [2^0, 2^(-1), 2^(-2), ..., 2^(1-NUM_PATCHMATCHNET_STAGES)]
+            #   Initially the intrinsics is scaled to fit the smallest image size
+            intrinsics[:2, :] /= 2 ** NUM_PATCHMATCHNET_STAGES
+
+            for j in range(self._num_stages):
```

Comment:
looks great @codyly, thanks for cleaning this all up.

One request here -- let's not index the stages by `j` which is a privileged variable we use to index tracks.

Let's call index it by `s`

Reply:
I just added a new commit to solve it. Thanks!